### PR TITLE
Add virtual thread support for JVMTI GetObjectMonitorUsage

### PR DESCRIPTION
**Pull request by babsingh**
- RI has removed support for virtual threads if an object monitor is
owned by a virtual thread.
- Support has also been removed for virtual threads that are waiting
be notified or own a monitor.
- Also, new tests show that waiting to be notified threads do not
repeat and only appear in one list. This has been corrected in OpenJ9.
- Update formatting in jvmtiObject.c to follow coding standards.

Fixes: #19497

### CODE REVIEW

**Comment by babsingh** [comment]
@ChengJin01 Would you like to review these changes?

fyi @hangshao0

File: runtime/jvmti/jvmtiObject.c [hunk 1]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
-							pStats->waiting[pStats->waitingCntr] =
-									(jthread) vm->internalVMFunctions->j9jni_createLocalRef(jniEnv, target);
-							++(pStats->waitingCntr);
-						}
+		J9JavaVM *vm = pStats->vm;
+		JNIEnv *jniEnv = (JNIEnv *)pStats->currentThread;
+
+		threadState &= ~(J9VMTHREAD_STATE_SUSPENDED | J9VMTHREAD_STATE_INTERRUPTED);
+		switch (threadState) {
+			case J9VMTHREAD_STATE_WAITING:
```
Comment (ChengJin01) [inline runtime/jvmti/jvmtiObject.c:239]: Please add something like `/* FALL THROUGH */` for the case.

File: runtime/jvmti/jvmtiObject.c [hunk 2]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
+			case J9VMTHREAD_STATE_WAITING_TIMED:
+				if (pStats->waiting == NULL) {
```
Comment (ChengJin01) [inline runtime/jvmti/jvmtiObject.c:241]: It would be `if (NULL == pStats->waiting) {`.

File: runtime/jvmti/jvmtiObject.c [hunk 3]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
-						if (pStats->blockedCntr < pStats->numBlocked) {
-							pStats->blocked[pStats->blockedCntr] =
-									(jthread) vm->internalVMFunctions->j9jni_createLocalRef(jniEnv, target);
-							++(pStats->blockedCntr);
-						}
+				}
+				break;
+
+			case J9VMTHREAD_STATE_BLOCKED:
+				if (pStats->blocked == NULL) {
```
Comment (ChengJin01) [inline runtime/jvmti/jvmtiObject.c:253]: It would be `if (NULL == pStats->blocked) {`.

File: runtime/jvmti/jvmtiObject.c [hunk 4]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
+					(pStats->numBlocked)++;
```
Comment (ChengJin01) [inline runtime/jvmti/jvmtiObject.c:254]: Should we change to `(pStats->numBlocked) += 1;`?

File: runtime/jvmti/jvmtiObject.c [hunk 5]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
+				} else {
+					if (pStats->blockedCntr < pStats->numBlocked) {
+						pStats->blocked[pStats->blockedCntr] =
+								(jthread) vm->internalVMFunctions->j9jni_createLocalRef(jniEnv, target);
+						++(pStats->blockedCntr);
```
Comment (ChengJin01) [inline runtime/jvmti/jvmtiObject.c:259]: Should we change to `(pStats->blockedCntr) += 1;`?

**Review by ChengJin01 — CHANGES_REQUESTED** [review]

**Comment by ChengJin01** [comment]
Add @fengxue-IS for review given this is a JVMTI specific PR.

File: runtime/jvmti/jvmtiObject.c (hunk 1 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c:239]: As per the [coding standard](https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md), `an exception to the fall-through rule is permitted for labels with no code`.

**Comment by babsingh** [comment]
@ChengJin01 I have address all of your feedback.

File: runtime/jvmti/jvmtiObject.c (hunk 5 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c]: Updated.

File: runtime/jvmti/jvmtiObject.c (hunk 4 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c]: Updated.

File: runtime/jvmti/jvmtiObject.c (hunk 3 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c]: Updated.

File: runtime/jvmti/jvmtiObject.c (hunk 2 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c]: Updated.

**Review by ChengJin01 — APPROVED** [review]
LGTM.

File: runtime/jvmti/jvmtiObject.c [hunk 6]
```diff
@@ -206,59 +209,60 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
-
-		j9object_t target = (j9object_t)vmThread->threadObject;
-
-		if (target) {
-
-			J9JavaVM * vm = pStats->vm;
-			JNIEnv * jniEnv = (JNIEnv *)pStats->currentThread;
-
-			/* CMVC 87023 - the spec is unclear, but from experimentation it appears that 'waiting to be notified' threads
-			 * should appear in both lists
```
Comment (fengxue-IS) [inline runtime/jvmti/jvmtiObject.c:228]: Is this comment still relevant? Base on the JDK bug reports[[1]](https://bugs.openjdk.org/browse/JDK-8247972)&[[2]](https://bugs.openjdk.org/browse/JDK-8328083), this issue is only fixed for 23+. 
There is no backport to earlier versions and given that the problem is a mismatch between intended spec and actual implementation, Im not sure if we should have this affect the existing behaviour for JDK 8/11/17.

Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c:228]: In their [FIX](https://github.com/openjdk/jdk/pull/17680), they updated existing tests and mentioned of dependencies in the certification tests. If we propagate this fix to JDK22-, then existing tests might fail. We should enable these changes only in JDK23+, and follow the RI as they backport the fix to older JDK versions.

Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c:228]: Updated changes to only apply to JDK23+.

**Review by fengxue-IS — APPROVED** [review]

**Comment by hangshao0** [comment]
jenkins test sanity,extended amac jdk23

**Comment by hangshao0** [comment]
jenkins test sanity.openjdk,extended.openjdk alinux jdk23

**Comment by babsingh** [comment]
Only known/unrelated failures are seen in the failing PR build.

File: runtime/jvmti/jvmtiObject.c [hunk 7]
```diff
@@ -206,59 +209,72 @@ jvmtiGetObjectMonitorUsage(jvmtiEnv* env,
...
 }
 
-
 static void
-findMonitorThreads(J9VMThread * vmThread, J9JVMTIMonitorStats * pStats)
+findMonitorThreads(J9VMThread *vmThread, J9JVMTIMonitorStats *pStats)
 {
-	j9object_t lockObject;
+	j9object_t lockObject = NULL;
+	J9VMThread *currentThread = (J9VMThread *)pStats->currentThread;
```
Comment (keithc-ca) [inline runtime/jvmti/jvmtiObject.c:216]: This needs to be conditional (Java 23+):
```
[2024-07-22T15:26:57.125Z] /Users/jenkins/workspace/Build_JDK21_aarch64_mac_Personal/openj9/runtime/jvmti/jvmtiObject.c:216:14: error: unused variable 'currentThread' [-Werror,-Wunused-variable]
[2024-07-22T15:26:57.125Z]         J9VMThread *currentThread = (J9VMThread *)pStats->currentThread;
[2024-07-22T15:26:57.125Z]                     ^
[2024-07-22T15:26:57.125Z] 1 error generated.
[2024-07-22T15:26:57.125Z] make[6]: *** [runtime/jvmti/CMakeFiles/j9jvmti.dir/jvmtiObject.c.o] Error 1
```

See https://openj9-jenkins.osuosl.org/job/Build_JDK21_aarch64_mac_Personal/154.

Comment (babsingh) [reply runtime/jvmti/jvmtiObject.c:216]: Opened https://github.com/eclipse-openj9/openj9/pull/19896 to address the above issue.

**Pull request merged**